File: HactoberFest/Java/Backtracking/Rat_In_a_Maze.java
public class RatMaze {


    public static boolean isSafe(int i, int j, int maze[][], int n){
        
       if((i >= 0 && i < n) && (j >= 0 && j<n) && maze[i][j] == 1) 
            return true;

       return false;
    }

    public static boolean solve(int i, int j, int maze[][], int n, int ans[][]){
        if(i == n-1 && j == n-1 && maze[i][j] == 1){
            ans[i][j] = 1;
            return true;
        }

        if(isSafe(i, j, maze, n)){
            if(ans[i][j] == 1)
                return false;
            ans[i][j] = 1;

            if(solve(i+1, j, maze, n, ans))
                return true;
            if(solve(i, j+1, maze, n, ans))
                return true;
            ans[i][j] = 0;
            return false;
        }

        return false;

    }
    public static void main(String[] args) {
        

        int maze[][] = { { 1, 0, 0, 0 },
                         { 1, 1, 0, 1 },
                         { 0, 1, 0, 0 },
                         { 1, 1, 1, 1 } };
        int n = maze.length;

        int ans[][] = new int[n][n];

        System.out.println(solve(0, 0, maze, n, ans));
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++)
                System.out.print(ans[i][j]);
            System.out.println();
        }



        

    }
}

File: Hacktoberfest-2023/DSA/Backtracking/Rat_in_a_maze.cpp
#include <iostream>
#include <vector>

using namespace std;

const int N = 4; // The size of the maze (N x N)

// Function to print the solution matrix
void printSolution(const vector<vector<int>>& solution) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            cout << solution[i][j] << " ";
        }
        cout << endl;
    }
}

// Function to check if the current cell is a valid move
bool isSafe(const vector<vector<int>>& maze, int x, int y) {
    return (x >= 0 && x < N && y >= 0 && y < N && maze[x][y] == 1);
}

// Recursive function to solve the Rat in a Maze problem
bool solveMazeUtil(const vector<vector<int>>& maze, int x, int y, vector<vector<int>>& solution) {
    // If the rat reaches the exit, we have a solution
    if (x == N - 1 && y == N - 1) {
        solution[x][y] = 1;
        return true;
    }

    // Check if the current cell is a valid move
    if (isSafe(maze, x, y)) {
        // Mark the cell as part of the solution path
        solution[x][y] = 1;

        // Move right
        if (solveMazeUtil(maze, x, y + 1, solution))
            return true;

        // Move down
        if (solveMazeUtil(maze, x + 1, y, solution))
            return true;

        // If no valid move is found, backtrack and mark the cell as not part of the solution path
        solution[x][y] = 0;
    }

    return false;
}

// Main function to solve the Rat in a Maze problem
void solveMaze(const vector<vector<int>>& maze) {
    vector<vector<int>> solution(N, vector<int>(N, 0)); // Create a solution matrix

    if (solveMazeUtil(maze, 0, 0, solution)) {
        cout << "Solution exists. Path taken by the rat:" << endl;
        printSolution(solution);
    } else {
        cout << "No solution exists." << endl;
    }
}

int main() {
    vector<vector<int>> maze = {
        {1, 0, 0, 0},
        {1, 1, 0, 1},
        {0, 1, 0, 0},
        {1, 1, 1, 1}
    };

    solveMaze(maze);

    return 0;
}
